python/pdo/contract/replication.py

```diff
@@ -0,0 +1,214 @@
+    
+    # -----------------------------------------------------------------
+
+    def __init__(self, replication_params, storage_clients, \
+                contract_id, blocks_to_replicate, enclave_service, data_dir=None):
+        """ Create an instance of the Replicator class: used for replicating the current change set"""
+        
+        self.storage_clients_to_replicate = []
+
+        if len(storage_clients) == 0:
+            return
+        
+        self.sservice_url_for_contract = enclave_service.storage_service_url
+        self.num_provable_replicas = replication_params['num_provable_replicas']
+        self.availability_duration = replication_params['availability_duration']
+        self.contract_id = contract_id
+        self.blocks_to_replicate = blocks_to_replicate
+        self.data_dir = data_dir
+
+        # identify reliable storage services to be used with this instance of Replicator
+        Replicator.__update_set_of_unreliable_storage_services__()
+        for client in  storage_clients:
+            use_client = True
+            if are_the_urls_same(client.ServiceURL, self.sservice_url_for_contract):
+                use_client = False
+            else:
+                for url in Replicator.__urls_of_storage_services_to_ignore__:
+                    if are_the_urls_same(client.ServiceURL, url):
+                        use_client = False
+                        break
+            if use_client:
+                self.storage_clients_to_replicate.append(client)
+         
+        # make sure that the number of reliable storage services is sufficient for replication 
+        if len(self.storage_clients_to_replicate) < replication_params['num_provable_replicas']-1: # -1 since we do not explicitly replicate to 
+                    # the sservice associated with the eservice
+            logger.exception('Cannot create Replicator instance: Insufficient number of reliable storage services available for replication')
+            raise ReplicationException('Cannot create Replicator instance: Insufficient number of reliable storage services available for replication')
+    
+    # -----------------------------------------------------------------
+
+    def push_blocks_to_storage_service(self, client, block_data_list, expiration):
+        """Wrapper around client.store_blocks to handle Exceptions during store_blocks. 
+        This is the task function used by a worker-thread that attempts replication to a specific storage service"""
+        
+        try:
+            response = client.store_blocks(block_data_list, expiration)
+            if response is None :
+                Replicator.__queue_of_SSURLS_raising_exceptions__.put(client.ServiceURL) 
+                logger.warn("Adding Storage Service at %s to the set of  bad ones : block_store_list is None", str(client.ServiceURL))
+        except Exception as e:
+            Replicator.__queue_of_SSURLS_raising_exceptions__.put(client.ServiceURL)
```

Comment:
in light of the comment above, we might at least pass the exception to the queue so one could do eventually a (un)garbage collection of the queue based on some exception analysis? Along this lines, adding a time-stamp could be useful also?

Reply:
Once again, as above, the exception queue n is a place holder for now.  Currently, the exception queue primarily helps to exit the program if we see an exception in another thread. A lot of how the exception must be handled belongs to the application. (a commi_id & a "logical" time stamp like you say could be added to the exception to help the application). 

And I agree with you, one could also do performance optimization based on exception: like retry a commit based on the exception,  report partial success (ex, replication succeeded, but transaction failed ), etc.  I am happy to discuss if there is non-trivial performance to be gained by exception analysis and make the necessary changes in a separate PR